### PR TITLE
ARROW-1199: [C++] Implement mutable POD struct for Array data

### PR DESCRIPTION
**Pull request by wesm**
This data structure provides a new internal data structure that is a self-contained representation of the memory and metadata inside an Arrow array data structure. 

This class is designed for easy internal data manipulation, analytical data processing, and data transport to and from IPC messages. For example, we could cast from int64 to float64 like so:

```c++
Int64Array arr = GetMyData();
std::shared_ptr<internal::ArrayData> new_data = arr->data()->ShallowCopy();
new_data->type = arrow::float64();
Float64Array double_arr(new_data);
```

This object is also useful in an analytics setting where memory may be reused. For example, if we had a group of operations all returning doubles, say:

```
Log(Sqrt(Expr(arr))
```

Then the low-level implementations of each of these functions could have the signatures

void Log(const ArrayData& values, ArrayData* out);

As another example a function may consume one or more memory buffers in an input array and replace them with newly-allocated data, changing the output data type as well.

I did quite a bit of refactoring and code simplification that was enabled by this patch. I note that performance in IPC loading of very wide record batches is about 15% slower, but in smaller record batches it is about the same in microbenchmarks. This code path could possibly be made faster with some performance analysis work.

### CODE REVIEW

**Comment by wesm** [comment]
I refactored to use the shared_ptr approach, which has a lot less copying, and will be better for nested data. Also changed internals of RecordBatch to be based on this simpler data structure. Performance is about the same it seems:

```
$ ./release/ipc-read-write-benchmark 
Run on (8 X 3670.51 MHz CPU s)
2017-07-10 04:12:36
Benchmark                                                  Time           CPU Iterations
----------------------------------------------------------------------------------------
BM_WriteRecordBatch/1/min_time:1.000/real_time         46326 ns      46318 ns      29789   21.0802GB/s
BM_WriteRecordBatch/4/min_time:1.000/real_time         47972 ns      47962 ns      28714   20.3569GB/s
BM_WriteRecordBatch/16/min_time:1.000/real_time        44827 ns      44815 ns      31817    21.785GB/s
BM_WriteRecordBatch/64/min_time:1.000/real_time        58261 ns      58248 ns      25366   16.7619GB/s
BM_WriteRecordBatch/256/min_time:1.000/real_time      116146 ns     116127 ns      12305   8.40806GB/s
BM_WriteRecordBatch/1024/min_time:1.000/real_time     382624 ns     382551 ns       4006   2.55228GB/s
BM_WriteRecordBatch/4k/min_time:1.000/real_time      1450150 ns    1449750 ns       1011   689.584MB/s
BM_WriteRecordBatch/8k/min_time:1.000/real_time      2790335 ns    2789340 ns        426    358.38MB/s
BM_ReadRecordBatch/1/min_time:1.000/real_time           1727 ns       1726 ns     801162   565.495GB/s
BM_ReadRecordBatch/4/min_time:1.000/real_time           3625 ns       3624 ns     386808   269.367GB/s
BM_ReadRecordBatch/16/min_time:1.000/real_time         10840 ns      10837 ns     125800   90.0847GB/s
BM_ReadRecordBatch/64/min_time:1.000/real_time         39714 ns      39709 ns      35407   24.5901GB/s
BM_ReadRecordBatch/256/min_time:1.000/real_time       232941 ns     232885 ns       5995   4.19232GB/s
BM_ReadRecordBatch/1024/min_time:1.000/real_time     1019862 ns    1019622 ns       1343   980.524MB/s
BM_ReadRecordBatch/4k/min_time:1.000/real_time       4185589 ns    4184750 ns        333   238.915MB/s
BM_ReadRecordBatch/8k/min_time:1.000/real_time       8630499 ns    8628216 ns        160   115.868MB/s
```

Probably need to make a FlameGraph to improve this any further

**Comment by wesm** [comment]
The Parquet build is failing due to some minor API renaming for consistency (data -> values). If you could also review the names for things hopefully we can avoid too many future API breakages like this

File: cpp/src/arrow/python/pandas_convert.cc [hunk 1]
```diff
@@ -1364,7 +1360,7 @@ inline Status ConvertStruct(const ChunkedArray& data, PyObject** out_values) {
   // ChunkedArray has at least one chunk
   auto arr = static_cast<const StructArray*>(data.chunk(0).get());
   // Use it to cache the struct type and number of fields for all chunks
-  auto num_fields = arr->fields().size();
+  int num_fields = arr->num_fields();
```
Comment (xhochy) [inline cpp/src/arrow/python/pandas_convert.cc]: `int32_t`

File: cpp/src/arrow/array.cc [hunk 2]
```diff
@@ -34,36 +34,19 @@
...
-      null_bitmap_data_(nullptr) {
-  if (null_count_ == 0) { null_bitmap_ = nullptr; }
-  if (null_bitmap_) { null_bitmap_data_ = null_bitmap_->data(); }
-}
-
 int64_t Array::null_count() const {
-  if (null_count_ < 0) {
-    if (null_bitmap_) {
-      null_count_ = length_ - CountSetBits(null_bitmap_data_, offset_, length_);
+  if (data_->null_count < 0) {
```
Comment (xhochy) [inline cpp/src/arrow/array.cc]: Take the chance and add an `UNLIKELY` ;)

**Review by xhochy — CHANGES_REQUESTED** [review]
@wesm I fail to see the need for it. Please take this as an effort to write some documentation that it even will be understandable for non-Arrow core developers ;)

Otherwise, I still reviewed the code fully. So once I'm convinced that this is needed all comments are the only outstanding issues.

File: cpp/src/arrow/array.h [hunk 3]
```diff
@@ -35,13 +35,88 @@
...
 class Status;
 
 template <typename T>
 struct Decimal;
 
+// ----------------------------------------------------------------------
+// Generic array data container
+
+/// \brief Container for generic array data
+struct ARROW_EXPORT ArrayData {
```
Comment (wesm) [inline cpp/src/arrow/array.h]: Done (so more explanation for why this is useful is there). I also moved this to an `arrow::internal` namespace and made it clear that this is for internal use and developers of Arrow. Should the need arise to make this a public API we can revisit

File: cpp/src/arrow/ipc/json-internal.cc [hunk 4]
```diff
@@ -558,7 +558,11 @@ class ArrayWriter {
   Status Visit(const StructArray& array) {
     WriteValidityField(array);
     const auto& type = static_cast<const StructType&>(*array.type());
-    return WriteChildren(type.children(), array.fields());
+    std::vector<std::shared_ptr<Array>> children;
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: `StructArray` is no longer storing the `arrow::Array` instances internally; they are constructed on demand by `arr->field(i)`. I am not sure we need to expose the vector of shared_ptr even if it existed

File: cpp/src/arrow/ipc/json-internal.cc [hunk 5]
```diff
@@ -569,7 +573,11 @@ class ArrayWriter {
     if (type.mode() == UnionMode::DENSE) {
       WriteIntegerField("OFFSET", array.raw_value_offsets(), array.length());
     }
-    return WriteChildren(type.children(), array.children());
+    std::vector<std::shared_ptr<Array>> children;
+    for (int i = 0; i < array.num_fields(); ++i) {
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: done

File: cpp/src/arrow/pretty_print.cc [hunk 6]
```diff
@@ -187,7 +187,11 @@ class ArrayPrinter {
 
   Status Visit(const StructArray& array) {
     RETURN_NOT_OK(WriteValidityBitmap(array));
-    return PrintChildren(array.fields(), array.offset(), array.length());
+    std::vector<std::shared_ptr<Array>> children;
```
Comment (wesm) [inline cpp/src/arrow/pretty_print.cc]: done

File: cpp/src/arrow/pretty_print.cc [hunk 7]
```diff
@@ -207,7 +211,11 @@ class ArrayPrinter {
     }
 
     // Print the children without any offset, because the type ids are absolute
-    return PrintChildren(array.children(), 0, array.length() + array.offset());
+    std::vector<std::shared_ptr<Array>> children;
```
Comment (wesm) [inline cpp/src/arrow/pretty_print.cc]: done

File: cpp/src/arrow/table.cc [hunk 8]
```diff
@@ -146,12 +146,23 @@ void AssertBatchValid(const RecordBatch& batch) {
 
 RecordBatch::RecordBatch(const std::shared_ptr<Schema>& schema, int64_t num_rows,
     const std::vector<std::shared_ptr<Array>>& columns)
-    : schema_(schema), num_rows_(num_rows), columns_(columns) {}
+    : schema_(schema), num_rows_(num_rows) {
+  columns_.reserve(columns.size());
```
Comment (wesm) [inline cpp/src/arrow/table.cc]: Done

File: cpp/src/arrow/table.h [hunk 9]
```diff
@@ -113,7 +114,7 @@ class ARROW_EXPORT RecordBatch {
       const std::vector<std::shared_ptr<Array>>& columns);
 
   RecordBatch(const std::shared_ptr<Schema>& schema, int64_t num_rows,
```
Comment (wesm) [inline cpp/src/arrow/table.h]: The move-based constructor was only being used internally AFAIK, but I have restored it

File: cpp/src/arrow/python/pandas_convert.cc (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: done

File: cpp/src/arrow/array.cc (hunk 2 above)
Comment (wesm) [inline cpp/src/arrow/array.cc]: done

**Comment by wesm** [comment]
Thanks for the review. Another benefit of this work is that the IPC loading machinery becomes substantially simpler. I was able to delete about half of array/loader.h and array/loader.cc and consolidate that code in arrow/ipc which makes more sense

**Review by xhochy — APPROVED** [review]
+1, merged once the build is fixed.

**Comment by wesm** [comment]
Got to get Python working again (this refactor has actually exposed a number of pre-existing bugs). Then will move onto GLib

**Comment by wesm** [comment]
OK, I'm going to give Windows a kick locally to see if any issues then merge this, then the Parquet patch, and will circle back if builds are failing

**Comment by wesm** [comment]
@xhochy I discovered a bug in ArrayBuilder::UnsafeSetNotNull https://github.com/apache/arrow/pull/824/commits/d0a8ee2bc4b3261976e6303c81ed1801bb869dcd; I'm amazed this one evaded us but it suggests our unit testing of the builders is not quite adequate

**Comment by wesm** [comment]
Merging this so we can try to get a passing Parquet build

**Pull request closed without merging**